Add ability to read MIE4NITF files

m-chaturvedi opened the pull request:
To read an MIE4NITF file, we do the following:

1. In the project file, say `test_project.prj`, we specify the absolute path to an image list file.
2. In the image list file, which is assumed to be in the same directory as the frames, we indicate MIE4NITF file using the prefix `MIE4NITF:`.

For example:

```
$ cat test_project.prj  
DataSetSpecifier=/home/chaturvedi/images/all_images.txt

-----

$ cat /home/chaturvedi/images/all_images.txt
frame__071.ntf
frame__072.ntf
frame__073.ntf
frame__074.ntf
frame__075.ntf
MIE4NITF:wpr-j2k.r0t0q0c1i1.NTF

-----

$ ls --format=single-column /home/chaturvedi/images/
all_images.txt
frame__001.ntf
frame__002.ntf
frame__071.ntf
frame__072.ntf
frame__073.ntf
frame__074.ntf
frame__075.ntf
wpr-j2k.r0t0q0c1i1.NTF
```

Note that `/home/chaturvedi/images/` may contain more files than what are used.

mwoehlke-kitware commented on Applications/VpView/vpFileDataSource.cxx:
```diff
@@ -128,6 +134,57 @@ void vpFileDataSource::setMonitoringEnabled(bool enable)
   }
 }
 
+bool add_mie4nitf_subdatasets (const QString &path, QStringList &list) {
```
Uh... your code style is rather all over the place. Please try to conform to the existing style. (n.b. `T& x`, not `T &x`, and [Allman braces](https://en.wikipedia.org/wiki/Indentation_style#Allman_style). Fortunately this is not one of the files still using historic VTK-style braces. Your indenting is also inconsistent.)

mwoehlke-kitware commented on Applications/VpView/vpFileDataSource.cxx:
```diff
@@ -128,6 +134,57 @@ void vpFileDataSource::setMonitoringEnabled(bool enable)
+
+    if (!dataset)
+    {
+      qWarning().nospace()<< "GDAL could not load file." <<
```
Why is `nospace()` needed here? Also, please put `<<` at the start of the next line, not the end of a line, and align with the prior `<<`.

mwoehlke-kitware commented on Applications/VpView/vpFileDataSource.cxx:
```diff
@@ -128,6 +134,57 @@ void vpFileDataSource::setMonitoringEnabled(bool enable)
+    if (!dataset)
+    {
+      qWarning().nospace()<< "GDAL could not load file." <<
+        path.toStdString().c_str();
```
Besides that this is the locale-incorrect way to convert to `std::string` (that would be [`stdString`](https://github.com/Kitware/qtextensions/blob/873c06769a2d76e2323152efc40ad910717ce648/core/qtStlUtil.h#L39)) or `char*` (that would be [`qPrintable`](https://doc.qt.io/qt-5/qtglobal.html#qPrintable)), why are you doing this at all? Just pass the `QString`...

mwoehlke-kitware commented on Applications/VpView/vpFileDataSource.cxx:
```diff
@@ -128,6 +134,57 @@ void vpFileDataSource::setMonitoringEnabled(bool enable)
+
+        const QRegExp rgx = QRegExp("SUBDATASET_\\d+_NAME");
+
+        if(key.contains(rgx))
```
Missing braces.

mwoehlke-kitware commented on Applications/VpView/vpFileDataSource.cxx:
```diff
@@ -154,21 +211,54 @@ void vpFileDataSource::update()
       while (!file.atEnd())
       {
-        const auto line = file.readLine();
+        QString line = file.readLine().trimmed();
```
Please don't decrease use of [AAA](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/).

mwoehlke-kitware commented on Applications/VpView/vpFileDataSource.cxx:
```diff
@@ -154,21 +211,54 @@ void vpFileDataSource::update()
+        // use glob to get more than one frames, we use the prefix `MIE4NITF:`
+        // to indicate that this file contains more than one frame in NITF
+        // format.
+        const QString mie4nitf_prefix = "MIE4NITF:";
```
Please use [`QStringLiteral`](https://doc.qt.io/qt-5/qstring.html#QStringLiteral).

mwoehlke-kitware commented on Applications/VpView/vpFileDataSource.cxx:
```diff
@@ -154,21 +211,54 @@ void vpFileDataSource::update()
+        // For ex., in MIE4NITF, frame # 3 inside `/A.ntf` (0-indexed) is
+        // stored in:
+        // `NITF_IM:2:/A.ntf`.
+        QString path_to_check;
```
Please use camelCase for variable names.

mwoehlke-kitware commented on Libraries/VtkVgIO/vtkGDALReader.cxx:
```diff
@@ -138,6 +138,7 @@ void vtkGDALReader::vtkGDALReaderInternal::ReadMetaData(
   this->ReleaseData();
 
   this->GDALData = (GDALDataset*) GDALOpen(fileName.c_str(), GA_ReadOnly);
+  std::cout << "Using GDAL reader to open file: " << fileName << std::endl;
```
I wonder how useful this is?

m-chaturvedi commented:
@mwoehlke-kitware, we have transferred the files to the following gitlab, I have addressed the comments there:
https://gitlab.kitware.com/LVMI/vivia/merge_requests/1/
